### PR TITLE
feat: async validate()

### PR DESCRIPTION
**Pull request by galvez**
Picking up on https://github.com/nuxt/nuxt.js/pull/3371

### CODE REVIEW

File: lib/app/client.js [hunk 1]
```diff
@@ -292,11 +292,22 @@ async function render (to, from, next) {
 
     // Call .validate()
     let isValid = true
-    Components.forEach((Component) => {
-      if (!isValid) return
```
Comment (galvez) [inline lib/app/client.js]: I have no idea why this `if()` was here ¯\_(ツ)_/¯

File: lib/app/client.js [hunk 2]
```diff
@@ -292,11 +292,22 @@ async function render (to, from, next) {
...
-      if (typeof Component.options.validate !== 'function') return
+    let validateReturn
+    for (const Component of Components) {
+      if (typeof Component.options.validate !== 'function') {
```
Comment (clarkdo) [inline lib/app/client.js]: Is it necessary to check？

File: lib/app/client.js [hunk 3]
```diff
@@ -292,11 +292,22 @@ async function render (to, from, next) {
...
+        continue
+      }
+
       isValid = Component.options.validate(app.context)
-    })
+      validateReturn = await Component.options.validate(app.context)
```
Comment (clarkdo) [inline lib/app/client.js]: I think await is fine, return value shouldn’t be nested promise

**Review by clarkdo — CHANGES_REQUESTED** [review]

**Comment by clarkdo** [comment]
Previous pull request is eligible to merge except unit test😄

**Comment by galvez** [comment]
> Previous pull request is eligible to merge except unit test😄

Yeah because it only implemented the server version :)

File: lib/app/client.js (hunk 3 above)
Comment (clarkdo) [reply lib/app/client.js]: And await works fine with non-async method

Comment (galvez) [reply lib/app/client.js]: > I think await is fine, return value shouldn’t be nested promise

The goal here is to support functions, async functions and functions that return promise (which are essentially the same as async functions) the same way. If a promise is returned, we need to await on it, if not, we can use the return value as is.

Comment (clarkdo) [reply lib/app/client.js]: They all work with await I think

Comment (clarkdo) [reply lib/app/client.js]: Return promise is same as async function.
Non async function will aotu append promise.resolve on return value if use await

Comment (galvez) [reply lib/app/client.js]: @clarkdo cool, I've updated it.

Comment (clarkdo) [reply lib/app/client.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#Description

**Review by clarkdo — APPROVED** [review]

**Review by pi0 — APPROVED** [review]
Nice

File: lib/app/server.js [hunk 4]
```diff
@@ -154,11 +154,18 @@ export default async (ssrContext) => {
   ** Call .validate()
   */
   let isValid = true
-  Components.forEach((Component) => {
-    if (!isValid) return
-    if (typeof Component.options.validate !== 'function') return
-    isValid = Component.options.validate(app.context)
-  })
+  let validateReturn
```
Comment (pi0) [inline lib/app/server.js]: Where is this being used??

File: lib/app/client.js [hunk 5]
```diff
@@ -292,11 +292,28 @@ async function render (to, from, next) {
 
     // Call .validate()
     let isValid = true
-    Components.forEach((Component) => {
-      if (!isValid) return
-      if (typeof Component.options.validate !== 'function') return
-      isValid = Component.options.validate(app.context)
-    })
+    let validationError
```
Comment (clarkdo) [inline lib/app/client.js]: Can be removed

**Pull request merged**